common/network-common/src/main/java/org/apache/spark/network/sasl/SparkSaslServer.java

```diff
@@ -150,6 +161,62 @@ public synchronized void dispose() {
   }
 
   /**
+   * Negotiate with peer for extended options, such as using AES cipher.
+   * @param message is message receive from peer which may contains communication parameters.
+   * @param callback is rpc callback.
+   * @param conf contains transport configuration.
+   * @return Object which represent the result of negotiate.
+   */
+  public Object negotiate(ByteBuffer message, RpcResponseCallback callback, TransportConf conf)
+    throws SaslException {
+    AesCipher cipher;
+
+    // Receive initial option from client
+    AesCipherOption cipherOption = AesCipherOption.decode(Unpooled.wrappedBuffer(message));
+    String transformation = AesCipher.TRANSFORM;
+    Properties properties = new Properties();
+
+    try {
+      // Generate key and iv
+      if (conf.saslEncryptionAesCipherKeySizeBits() % 8 != 0) {
+        throw new IllegalArgumentException("The AES cipher key size in bits should be a multiple " +
+          "of byte");
+      }
+
+      int keyLen = conf.saslEncryptionAesCipherKeySizeBits() / 8;
+      int paramLen = CryptoCipherFactory.getCryptoCipher(transformation,properties).getBlockSize();
+      byte[] inKey = new byte[keyLen];
+      byte[] outKey = new byte[keyLen];
+      byte[] inIv = new byte[paramLen];
+      byte[] outIv = new byte[paramLen];
+
+      // Get the 'CryptoRandom' instance.
+      CryptoRandom random = CryptoRandomFactory.getCryptoRandom(properties);
+      random.nextBytes(inKey);
+      random.nextBytes(outKey);
+      random.nextBytes(inIv);
+      random.nextBytes(outIv);
+
+      // Update cipher option for client. The key is encrypted.
+      cipherOption.setParameters(wrap(inKey, 0, inKey.length), inIv,
+        wrap(outKey, 0, outKey.length), outIv);
+
+      // Enable AES on saslServer
+      cipher = new AesCipher(properties, inKey, outKey, inIv, outIv);
+
+      // Send cipher option to client
+      ByteBuf buf = Unpooled.buffer(cipherOption.encodedLength());
+      cipherOption.encode(buf);
+      callback.onSuccess(buf.nioBuffer());
+    } catch (Exception e) {
+      logger.error("AES negotiation exception: ", e);
```

Comment:
nit: `...exception.` (the colon is not necessary and looks weird in the logged message)